src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyTests.cs

```diff
@@ -0,0 +1,1005 @@
+        }
+
+        [Fact]
+        public void NoAssignmentToReadOnlyParameterInAnonymousMethod()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    public C()
+    {
+        Action<int> a = delegate (readonly int i)
+        {
+            i = 0;
+        };
+    }
+}
+";
+
+            CreateCompilationWithMscorlib(source, parseOptions: options).VerifyDiagnostics(
+                // (10,13): error CS1656: Cannot assign to 'i' because it is a 'readonly parameter'
+                //             i = 0;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocalCause, "i").WithArguments("i", "readonly parameter").WithLocation(10, 13));
+        }
+
+        [Fact]
+        public void TestLocalFunction1()
+        {
+            var source = @"
+class C
+{
+    public C()
+    {
+        void LocalFunction(readonly int i)
+        {
+        }
+    }
+}
+";
+
+            CreateCompilationWithMscorlib(source, parseOptions: options).VerifyDiagnostics(
+                // (6,14): warning CS0168: The variable 'LocalFunction' is declared but never used
+                //         void LocalFunction(readonly int i)
+                Diagnostic(ErrorCode.WRN_UnreferencedVar, "LocalFunction").WithArguments("LocalFunction").WithLocation(6, 14));
+        }
+
+        [Fact]
+        public void TestLocalFunction2()
+        {
+            var source = @"
+class C
+{
+    public C()
+    {
+        void LocalFunction(readonly int i)
+        {
+            i = 0;
+        }
+    }
+}
+";
+
+            CreateCompilationWithMscorlib(source, parseOptions: options).VerifyDiagnostics(
+                // (8,13): error CS1656: Cannot assign to 'i' because it is a 'readonly parameter'
+                //             i = 0;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocalCause, "i").WithArguments("i", "readonly parameter").WithLocation(8, 13),
+                // (6,14): warning CS0168: The variable 'LocalFunction' is declared but never used
+                //         void LocalFunction(readonly int i)
+                Diagnostic(ErrorCode.WRN_UnreferencedVar, "LocalFunction").WithArguments("LocalFunction").WithLocation(6, 14));
+        }
+
+        [Fact]
+        public void TestReadOnlyForVariableDeclaration()
+        {
+            // TODO(cyrusn): It would be nice if we gave a better error message here, rather than
+            // the parser going off the rails.
+            //
+            // Note: i don't see any benefit in actually supporting 'readonly' here as why would
+            // you have a for loop that made its iteration variable readonly?
+            var source = @"
+class C
+{
+    public C()
+    {
+        for (readonly int i = 0;;) { }
+    }
+}
+";
+
+            CreateCompilationWithMscorlib(source, parseOptions: options).VerifyDiagnostics(
+                // (6,14): error CS0769: 'readonly' is not valid with 'for' variables. 'for' variables are already readonly
```

Comment:
> 'for' variables are already readonly
This is incorrect, 'for' variables are not readonly. Though readonly 'for' variable will have very little value. The error should be different I think.